File: algorithmic-toolbox/src/week1/java/PartialSumFibonacci.java
package week1.java;

import java.util.Scanner;

class PartialSumFibonacci {

    public static void main(final String[] args) {
        try (final Scanner in = new Scanner(System.in)) {
            System.out.println(partialSumFib(in.nextLong(), in.nextLong()));
        }
    }

    private static long partialSumFib(final long start, final long end) {
        final long prefix = sumFib(start-1), suffix = sumFib(end);
        return (suffix < prefix) ? (suffix + 10) - prefix : suffix - prefix;
    }

    private static long sumFib(final long n) {
        final int[][] startMatrix = {{1, 1, 0}, {1, 0, 0}, {1, 0, 1}};
        final int[][] resultMatrix = pow(startMatrix, n+1);
        return resultMatrix[2][1];
    }

    private static int[][] pow(final int[][] matrix, final long n) {
        if (n <= 1) {
            return matrix;
        } else if (n == 2) {
            return square(matrix);
        } else if ((n % 2) == 0) {
            return pow(square(matrix), n/2);
        } else {
            return mult(matrix, pow(square(matrix), (n-1)/2));
        }
    }

    private static int[][] square(final int[][] matrix) {
        return mult(matrix, matrix);
    }

    private static int[][] mult(final int[][] a, final int[][] b) {
        return new int[][] {
            { m(m(a[0][0]*b[0][0])+m(a[0][1]*b[1][0])+m(a[0][2]*b[2][0])), m(m(a[0][0]*b[0][1])+m(a[0][1]*b[1][1])+m(a[0][2]*b[2][1])), m(m(a[0][0]*b[0][2])+m(a[0][1]*b[1][2])+m(a[0][2]*b[2][2])) },
            { m(m(a[1][0]*b[0][0])+m(a[1][1]*b[1][0])+m(a[1][2]*b[2][0])), m(m(a[1][0]*b[0][1])+m(a[1][1]*b[1][1])+m(a[1][2]*b[2][1])), m(m(a[1][0]*b[0][2])+m(a[1][1]*b[1][2])+m(a[1][2]*b[2][2])) },
            { m(m(a[2][0]*b[0][0])+m(a[2][1]*b[1][0])+m(a[2][2]*b[2][0])), m(m(a[2][0]*b[0][1])+m(a[2][1]*b[1][1])+m(a[2][2]*b[2][1])), m(m(a[2][0]*b[0][2])+m(a[2][1]*b[1][2])+m(a[2][2]*b[2][2])) }
        };
    }

    private static int m(final int n) {
        return (n % 10);
    }
}

File: algorithmic-toolbox/src/week1/scala/PartialSumFibonacci.scala
object PartialSumFibonacci {

  def main(args: Array[String]) = {
    val in = new java.util.Scanner(System.in)
    println(partialSumFib(in.nextLong, in.nextLong))
  }

  def partialSumFib(start: Long, end: Long) = {
    type Matrix = ((Int, Int, Int), (Int, Int, Int), (Int, Int, Int))

    def sumFib(n: Long) = {
      def square(m: Matrix) = mult(m, m)
      def m(n: Int) = n % 10
      def mult(a: Matrix, b: Matrix) = (
        (
          m(m(a._1._1 * b._1._1) + m(a._1._2 * b._2._1) + m(a._1._3 * b._3._1)),
          m(m(a._1._1 * b._1._2) + m(a._1._2 * b._2._2) + m(a._1._3 * b._3._2)),
          m(m(a._1._1 * b._1._3) + m(a._1._2 * b._2._3) + m(a._1._3 * b._3._3))),
        (
          m(m(a._2._1 * b._1._1) + m(a._2._2 * b._2._1) + m(a._2._3 * b._3._1)),
          m(m(a._2._1 * b._1._2) + m(a._2._2 * b._2._2) + m(a._2._3 * b._3._2)),
          m(m(a._2._1 * b._1._3) + m(a._2._2 * b._2._3) + m(a._2._3 * b._3._3))),
        (
          m(m(a._3._1 * b._1._1) + m(a._3._2 * b._2._1) + m(a._3._3 * b._3._1)),
          m(m(a._3._1 * b._1._2) + m(a._3._2 * b._2._2) + m(a._3._3 * b._3._2)),
          m(m(a._3._1 * b._1._3) + m(a._3._2 * b._2._3) + m(a._3._3 * b._3._3))))
  
      def pow(m: Matrix, p: Long): Matrix = {
        if (p <= 1) m
        else if (p == 2) square(m)
        else if ((p % 2) == 0) pow(square(m), p / 2);
        else mult(m, pow(square(m), (p - 1) / 2));
      }
  
      pow(((1, 1, 0), (1, 0, 0), (1, 0, 1)), n + 1)._3._2
    }

    val (prefix, suffix) = (sumFib(start - 1), sumFib(end))
    if (suffix < prefix) (10 + suffix) - prefix
    else suffix - prefix
  }
}